### PR TITLE
Initial SQLite migration and structure

### PR DESCRIPTION
**Pull request by LukasPukenis**
This PR prepares the groundwork for SQLite3 and is part of upcoming PRs based on this one.

### CODE REVIEW

File: drop-storage/sqlx-data.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (matszczygiel) [inline drop-storage/sqlx-data.json:1]: Is this file necessary?

File: drop-storage/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+type Result<T> = std::result::Result<T, Error>;
+// SQLite storage wrapper
+pub struct Storage {
+    _logger: Logger,
+    #[allow(dead_code)]
+    conn: SqlitePool,
+}
+
+#[derive(Debug, strum::Display, strum::EnumString, PartialEq, Eq)]
```
Comment (matszczygiel) [inline drop-storage/src/lib.rs:18]: `strum` would probably be replaced with something like this https://github.com/launchbadge/sqlx/blob/main/examples/postgres/files/src/main.rs#L4

File: drop-storage/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        debug!(logger, "SQLite existing: {}", path);
+
+        let mut options = SqliteConnectOptions::from_str(path)?;
+        options.log_statements(log::LevelFilter::Debug);
+
+        let conn = SqlitePool::connect_with(options)
+            .await
+            .map_err(|e| error::Error::InternalError(e.to_string()))
```
Comment (matszczygiel) [inline drop-storage/src/lib.rs:93]: Duplicated line?

File: drop-config/src/lib.rs [hunk 4]
```diff
@@ -6,13 +6,14 @@ pub struct Config {
     pub moose: MooseConfig,
 }
 
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Clone)]
```
Comment (matszczygiel) [inline drop-config/src/lib.rs:9]: If we're dropping `Copy` I suggest wrapping config in `Arc` for cheap cloning in `drop-transfer` crate

File: drop-transfer/src/manager.rs [hunk 5]
```diff
@@ -30,9 +30,10 @@ pub struct TransferState {
 
 /// Transfer manager is responsible for keeping track of all ongoing or pending
 /// transfers and their status
-#[derive(Default)]
 pub(crate) struct TransferManager {
     transfers: HashMap<Uuid, TransferState>,
+    #[allow(dead_code)]
+    storage: Arc<Mutex<drop_storage::Storage>>,
```
Comment (matszczygiel) [inline drop-transfer/src/manager.rs:36]: Why do we use `Arc` here?

File: drop-storage/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+version = "1.0.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+thiserror = "*"
+sqlx = { version = "0.6", features = [ "runtime-tokio-rustls", "sqlite", "time", "chrono" ] }
+slog = { workspace = true }
+tokio = { version = "1", features = ["full"] }
```
Comment (matszczygiel) [inline drop-storage/Cargo.toml:12]:
```suggestion
tokio = { workspace = true, features = ["full"] }
```
Are you sure that you need all the features?

File: drop-storage/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+tokio-util = "0.7.3"
+tokio-test = "*"
+strum = { version = "0.24.1", features = ["derive"] }
+inspect = "0.1.2"
+async-std = { version = "1", features = ["attributes", "tokio1"] }
+log = "0.4"
```
Comment (matszczygiel) [inline drop-storage/Cargo.toml:18]: It's unused, besides we're using `slog` instead of `log` crate

File: drop-storage/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+[dependencies]
+thiserror = "*"
+sqlx = { version = "0.6", features = [ "runtime-tokio-rustls", "sqlite", "time", "chrono" ] }
+slog = { workspace = true }
+tokio = { version = "1", features = ["full"] }
+tokio-util = "0.7.3"
+tokio-test = "*"
+strum = { version = "0.24.1", features = ["derive"] }
+inspect = "0.1.2"
```
Comment (matszczygiel) [inline drop-storage/Cargo.toml:16]: Unused

File: drop-storage/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+async-std = { version = "1", features = ["attributes", "tokio1"] }
```
Comment (matszczygiel) [inline drop-storage/Cargo.toml:17]: Do we need this ?

File: drop-storage/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+thiserror = "*"
+sqlx = { version = "0.6", features = [ "runtime-tokio-rustls", "sqlite", "time", "chrono" ] }
+slog = { workspace = true }
+tokio = { version = "1", features = ["full"] }
+tokio-util = "0.7.3"
+tokio-test = "*"
```
Comment (matszczygiel) [inline drop-storage/Cargo.toml:14]: Do we need these two?

**Review by matszczygiel — CHANGES_REQUESTED** [review]

File: drop-storage/migrations/20230406110818_initial_migration.sql [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  is_outgoing INTEGER NOT NULL,
+
+  created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  FOREIGN KEY(peer_id) REFERENCES peers(id) ON DELETE CASCADE ON UPDATE CASCADE,
+  CHECK(is_outgoing = 0 OR is_outgoing = 1)
+);
+
+-- added paths to the outgoing transfer
+CREATE TABLE IF NOT EXISTS outgoing_paths (
+  id INTEGER PRIMARY KEY AUTOINCREMENT,
```
[the remark this answers is not in the document]
Comment (LukasPukenis) [reply drop-storage/migrations/20230406110818_initial_migration.sql:19]: `transfer_id` is used in this table as a foreign key.

File: drop-storage/sqlx-data.json (hunk 1 above)
Comment (LukasPukenis) [reply drop-storage/sqlx-data.json:1]: `sqlx` created it so I assumed it's somewhat needed but I can move it into gitignore I guess

File: drop-storage/src/lib.rs (hunk 3 above)
Comment (LukasPukenis) [reply drop-storage/src/lib.rs:93]: oh wow, really :D

File: drop-storage/migrations/20230406110818_initial_migration.sql (hunk 11 above)
[the remark this answers is not in the document]
Comment (matszczygiel) [reply drop-storage/migrations/20230406110818_initial_migration.sql:19]: Oh, sorry. My bad. I meant the `transfers` table but it's already indexed by uuid. NWM

File: drop-storage/sqlx-data.json (hunk 1 above)
Comment (matszczygiel) [reply drop-storage/sqlx-data.json:1]: Well, it's needed in the offline case. Which probably is irrelevant to the SQLite case. I would go for removing it

File: drop-storage/Cargo.toml (hunk 6 above)
Comment (LukasPukenis) [reply drop-storage/Cargo.toml]: this PR was ripped out of a stable branch, these were used in unit tests, however as this PR doesn't contain them, these are unused, nice catch

File: drop-storage/sqlx-data.json (hunk 1 above)
Comment (LukasPukenis) [reply drop-storage/sqlx-data.json]: moved to gitignore and removed it

File: drop-storage/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            debug!(logger, "SQLite not existing. Creating database at {}", path);
+            Sqlite::create_database(path)
+                .await
+                .map_err(|e| error::Error::InternalError(e.to_string()))?;
+        }
+
+        debug!(logger, "SQLite existing: {}", path);
+
+        let options = SqliteConnectOptions::from_str(path)?;
```
Comment (matszczygiel) [inline drop-storage/src/lib.rs:88]: The `SqliteConnectOptions` has a `create_if_missing` property. Can we use it to replace this logic?

File: drop-storage/src/lib.rs (hunk 2 above)
Comment (LukasPukenis) [reply drop-storage/src/lib.rs]: That's a nice suggestion, even though the structure is not really used in this MR and thus I totally removed it

File: drop-storage/src/lib.rs (hunk 12 above)
Comment (LukasPukenis) [reply drop-storage/src/lib.rs:30]: Oh that's nice, thanks

File: drop-storage/src/lib.rs (hunk 3 above)
Comment (matszczygiel) [reply drop-storage/src/lib.rs:23]: I still see it
image

Is this intentional ?

File: drop-config/src/lib.rs (hunk 4 above)
Comment (LukasPukenis) [reply drop-config/src/lib.rs:9]: nice suggestion

File: drop-transfer/src/transfer.rs [hunk 13]
```diff
@@ -16,15 +16,15 @@ pub struct Transfer {
...
+    pub fn new(peer: IpAddr, files: Vec<File>, config: Arc<DropConfig>) -> Result<Self> {
         Self::new_with_uuid(peer, files, Uuid::new_v4(), config)
     }
 
     pub(crate) fn new_with_uuid(
         peer: IpAddr,
         files: Vec<File>,
         uuid: Uuid,
-        config: &DropConfig,
+        config: Arc<DropConfig>,
```
Comment (matszczygiel) [inline drop-transfer/src/transfer.rs:27]: Um, why? We don't need ownership of config here, plain reference suffices

File: norddrop/src/device/mod.rs [hunk 14]
```diff
@@ -65,7 +65,7 @@ impl NordDropFFI {
                 cb: event_cb,
                 logger,
             }),
-            config: Config::default(),
+            config: Arc::new(Config::default()),
```
Comment (matszczygiel) [inline norddrop/src/device/mod.rs:68]: Probably storing `Arc` here is a huge pain. I would go for storing here as plain data here, and wrap it into arc inside `Service` structure

**Review by matszczygiel — APPROVED** [review]
I think there are many places where we can just pass `&DropConfig`. But this can be sorted out later

File: drop-transfer/src/transfer.rs (hunk 13 above)
Comment (LukasPukenis) [reply drop-transfer/src/transfer.rs]: I changed DropConfig to be wrapped inside an Arc and fixed a ton of issues surrounding that so it simply slipped here and here. Anyway removed it

File: norddrop/src/device/mod.rs (hunk 14 above)
Comment (LukasPukenis) [reply norddrop/src/device/mod.rs]: I think it's not that painful and has certain benefits:
- Easy not to think about lifetimes
- Passing the config and updating it in the future makes it just work

Currently, if I will have it as a plain member and wrap it into an Arc<> when passing, I am gonna pass clones instead thus making each transfer with an explicit clone of a config. This is not an issue and depends on the point of view. I changed it into the non-arc member for the piece of mind

**Review by matszczygiel — APPROVED** [review]

File: norddrop/src/device/mod.rs [hunk 15]
```diff
@@ -132,11 +133,20 @@ impl NordDropFFI {
...
+                        ffi::types::NORDDROP_RES_DB_ERROR
+                    })?;
+
             let service = match Service::start(
                 addr,
+                storage,
                 tx,
                 self.logger.clone(),
-                self.config.drop,
+                Arc::new(self.config.clone().drop),
```
[the remark this answers is not in the document]
Comment (LukasPukenis) [reply norddrop/src/device/mod.rs]: thanks

**Review by matszczygiel — APPROVED** [review]
:+1:
